File: web/src/features/editor/domain/document-v2.test.ts
import { readFileSync } from 'node:fs'
import { describe, expect, it } from 'vitest'
import type { EditorDocumentV1 } from './document'
import type { EditorDocumentV2 } from './document-v2'
import {
  compileEditorDocumentV2ToV1,
  migrateEditorDocumentV1ToV2,
  UnsupportedEditorSemanticsError,
  validateEditorDocumentV2,
} from './document-v2'

const fixture = (name: string) =>
  JSON.parse(
    readFileSync(
      new URL(`../../../../../testdata/editor/${name}`, import.meta.url),
      'utf8',
    ),
  )

describe('editor document V2', () => {
  it('matches the shared V1 migration golden and round trips the renderable subset', () => {
    const v1 = fixture('v1-flat.json') as EditorDocumentV1
    const expected = fixture('v2-flat.json') as EditorDocumentV2
    const v2 = migrateEditorDocumentV1ToV2(v1)
    expect(v2).toEqual(expected)
    expect(compileEditorDocumentV2ToV1(v2)).toEqual({
      ...v1,
      objects: [...v1.objects]
        .sort((a, b) => a.z_index - b.z_index)
        .map((object, index) => ({ ...object, z_index: index })),
    })
  })

  it('validates professional tree semantics without silently flattening them', () => {
    const v2 = fixture('v2-flat.json') as EditorDocumentV2
    v2.nodes.push({
      id: 'group',
      type: 'group',
      parent_id: null,
      order_key: '00000002',
      transform: [1, 0, 0, 1, 0, 0],
      opacity: 1,
      blend_mode: 'normal',
      visible: true,
      locked: false,
    })
    v2.nodes[0].parent_id = 'group'
    expect(validateEditorDocumentV2(v2)).toEqual([])
    expect(() => compileEditorDocumentV2ToV1(v2)).toThrow(
      UnsupportedEditorSemanticsError,
    )
  })

  it('validates an explicit clipped adjustment target', () => {
    const v2 = fixture('v2-flat.json') as EditorDocumentV2
    const target = v2.nodes[0]
    v2.nodes.push({
      id: 'adjustment',
      type: 'adjustment',
      target_id: target.id,
      parent_id: target.parent_id,
      order_key: '00000003',
      transform: [1, 0, 0, 1, 0, 0],
      opacity: 0.5,
      blend_mode: 'normal',
      visible: true,
      locked: false,
      effects: [],
    })
    expect(validateEditorDocumentV2(v2)).toEqual([])

    v2.nodes.at(-1)!.target_id = 'missing'
    expect(validateEditorDocumentV2(v2)).toContain('target:adjustment')
  })

  it('validates shape masks and refuses lossy V1 compilation', () => {
    const v2 = fixture('v2-flat.json') as EditorDocumentV2
    v2.nodes[0].shape_mask = {
      type: 'ellipse',
      x: 0.1,
      y: 0.2,
      width: 0.6,
      height: 0.5,
      inverted: true,
    }
    expect(validateEditorDocumentV2(v2)).toEqual([])
    expect(() => compileEditorDocumentV2ToV1(v2)).toThrow(
      UnsupportedEditorSemanticsError,
    )
    v2.nodes[0].crop = { x: 0, y: 0, width: 0.5, height: 0.5 }
    expect(validateEditorDocumentV2(v2)).toContain(`asset:${v2.nodes[0].id}`)
  })

  it('validates immutable pixel mask references', () => {
    const v2 = migrateEditorDocumentV1ToV2(
      fixture('v1-flat.json') as EditorDocumentV1,
    )
    v2.nodes[0].pixel_mask = {
      resource_id: '8cab813d-008a-49f2-bfe4-1e1d4ca45b60',
      version: 4,
    }
    expect(validateEditorDocumentV2(v2)).toEqual([])
    expect(() => compileEditorDocumentV2ToV1(v2)).toThrow(
      UnsupportedEditorSemanticsError,
    )
    v2.nodes[0].pixel_mask.version = -1
    expect(validateEditorDocumentV2(v2)).toContain(`asset:${v2.nodes[0].id}`)
  })
})

File: backend/internal/editor/document_v2_test.go
package editor

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"github.com/google/uuid"
)

func TestDocumentV2SharedGoldenMigration(t *testing.T) {
	v1Raw, err := os.ReadFile(filepath.Join("..", "..", "..", "testdata", "editor", "v1-flat.json"))
	if err != nil {
		t.Fatal(err)
	}
	v2Raw, err := os.ReadFile(filepath.Join("..", "..", "..", "testdata", "editor", "v2-flat.json"))
	if err != nil {
		t.Fatal(err)
	}
	v1, err := Decode(v1Raw)
	if err != nil {
		t.Fatal(err)
	}
	got, err := MigrateV1ToV2(v1)
	if err != nil {
		t.Fatal(err)
	}
	gotRaw, err := json.Marshal(got)
	if err != nil {
		t.Fatal(err)
	}
	var gotCanonical, wantCanonical bytes.Buffer
	if err = json.Compact(&gotCanonical, gotRaw); err != nil {
		t.Fatal(err)
	}
	if err = json.Compact(&wantCanonical, v2Raw); err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(gotCanonical.Bytes(), wantCanonical.Bytes()) {
		t.Fatalf("migration differs from shared golden\n got: %s\nwant: %s", gotCanonical.Bytes(), wantCanonical.Bytes())
	}
	decoded, err := DecodeAny(v2Raw)
	if err != nil || decoded.SchemaVersion != 2 || len(decoded.AssetIDs()) != 2 {
		t.Fatalf("DecodeAny(V2) = %#v, %v", decoded, err)
	}
}

func TestDecodeAnyPreservesPerVersionSizeLimits(t *testing.T) {
	oversizedV1 := append([]byte(`{"schema_version":1,"canvas":{"width":1,"height":1},"objects":[]}`), bytes.Repeat([]byte(" "), MaxDocumentBytes)...)
	if _, err := DecodeAny(oversizedV1); !errors.Is(err, ErrDocumentTooLarge) {
		t.Fatalf("oversized V1 error = %v", err)
	}
	v2, err := MigrateV1ToV2(New(uuid.New(), 32, 32))
	if err != nil {
		t.Fatal(err)
	}
	raw, err := json.Marshal(v2)
	if err != nil {
		t.Fatal(err)
	}
	if _, err = DecodeAny(raw); err != nil {
		t.Fatalf("valid V2 error = %v", err)
	}
}

func TestDocumentV2MigratesV1WithoutLosingRenderableSemantics(t *testing.T) {
	v1 := New(uuid.New(), 2048, 1024)
	v1.Objects[0].Crop = &Crop{X: 0.1, Y: 0.2, Width: 0.7, Height: 0.6}
	v1.Objects[0].Opacity = 0.75
	v2, err := MigrateV1ToV2(v1)
	if err != nil {
		t.Fatal(err)
	}
	if v2.SchemaVersion != 2 || v2.RendererSemanticsVersion != 1 || len(v2.Nodes) != 1 || v2.Nodes[0].Type != "raster" {
		t.Fatalf("unexpected V2 document: %#v", v2)
	}
	roundTrip, err := v2.ToV1()
	if err != nil {
		t.Fatal(err)
	}
	if roundTrip.Objects[0].AssetID != v1.Objects[0].AssetID || roundTrip.Objects[0].Opacity != v1.Objects[0].Opacity || *roundTrip.Objects[0].Crop != *v1.Objects[0].Crop {
		t.Fatalf("round trip changed semantics: %#v", roundTrip)
	}
}

func TestDocumentV2ValidatesTreeEffectsAndExportSubset(t *testing.T) {
	v2, err := MigrateV1ToV2(New(uuid.New(), 1024, 1024))
	if err != nil {
		t.Fatal(err)
	}
	groupID := "group"
	v2.Nodes = append(v2.Nodes, NodeV2{
		ID: groupID, Type: "group", Name: "组", OrderKey: "00000001",
		Transform: [6]float64{1, 0, 0, 1, 0, 0}, Opacity: 1, BlendMode: "normal", Visible: true,
	})
	v2.Nodes[0].ParentID = &groupID
	v2.Nodes[0].Effects = []EffectV2{{Type: "exposure", Version: 1, Enabled: true, Parameters: map[string]float64{"stops": 0.5}}}
	if err := v2.Validate(); err != nil {
		t.Fatalf("valid tree rejected: %v", err)
	}
	if _, err := v2.ToV1(); !errors.Is(err, ErrUnsupportedDocumentSemantics) {
		t.Fatalf("professional semantics export error = %v", err)
	}
	v2.Nodes[1].ParentID = &v2.Nodes[0].ID
	if err := v2.Validate(); !errors.Is(err, ErrInvalidDocument) {
		t.Fatalf("cycle error = %v", err)
	}
}

func TestDocumentV2ValidatesClippedAdjustmentTarget(t *testing.T) {
	v2, err := MigrateV1ToV2(New(uuid.New(), 128, 128))
	if err != nil {
		t.Fatal(err)
	}
	targetID := v2.Nodes[0].ID
	v2.Nodes = append(v2.Nodes, NodeV2{ID: "adjustment", Type: "adjustment", TargetID: &targetID, OrderKey: "00000002", Transform: [6]float64{1, 0, 0, 1, 0, 0}, Opacity: .5, BlendMode: "normal", Visible: true, Effects: []EffectV2{}})
	if err = v2.Validate(); err != nil {
		t.Fatalf("valid adjustment rejected: %v", err)
	}
	missing := "missing"
	v2.Nodes[1].TargetID = &missing
	if err = v2.Validate(); !errors.Is(err, ErrInvalidDocument) {
		t.Fatalf("missing target error = %v", err)
	}
}

func TestDocumentV2ValidatesShapeMaskWithoutSilentV1Downgrade(t *testing.T) {
	v2, err := MigrateV1ToV2(New(uuid.New(), 128, 128))
	if err != nil {
		t.Fatal(err)
	}
	v2.Nodes[0].ShapeMask = &ShapeMaskV2{Type: "ellipse", X: .1, Y: .2, Width: .6, Height: .5, Inverted: true}
	if err = v2.Validate(); err != nil {
		t.Fatalf("valid shape mask rejected: %v", err)
	}
	if _, err = v2.ToV1(); !errors.Is(err, ErrUnsupportedDocumentSemantics) {
		t.Fatalf("shape mask downgrade error = %v", err)
	}
	v2.Nodes[0].Crop = &Crop{X: 0, Y: 0, Width: .5, Height: .5}
	if err = v2.Validate(); !errors.Is(err, ErrInvalidDocument) {
		t.Fatalf("crop and shape mask conflict = %v", err)
	}
}

func TestDecodeV2RejectsUnknownFieldsAndExcessiveDepth(t *testing.T) {
	v2, err := MigrateV1ToV2(New(uuid.New(), 32, 32))
	if err != nil {
		t.Fatal(err)
	}
	raw, _ := json.Marshal(v2)
	raw[len(raw)-1] = ','
	raw = append(raw, []byte(`"url":"https://example.com"}`)...)
	if _, err := DecodeV2(raw); !errors.Is(err, ErrInvalidDocument) {
		t.Fatalf("unknown field error = %v", err)
	}
	deep := v2
	deep.Nodes = nil
	var parent *string
	for index := 0; index <= MaxNodeDepthV2; index++ {
		id := fmt.Sprintf("group-%02d", index)
		deep.Nodes = append(deep.Nodes, NodeV2{ID: id, Type: "group", ParentID: parent, OrderKey: fmt.Sprintf("%08d", index), Transform: [6]float64{1, 0, 0, 1, 0, 0}, Opacity: 1, BlendMode: "normal", Visible: true})
		parent = &deep.Nodes[len(deep.Nodes)-1].ID
	}
	if err := deep.Validate(); !errors.Is(err, ErrInvalidDocument) {
		t.Fatalf("depth error = %v", err)
	}
}
